api/rosetta/combine.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package rosetta
+
+import (
+	"encoding/json"
+	"errors"
+	"net/http"
+
+	"github.com/labstack/echo/v4"
+
+	"github.com/onflow/flow-go-sdk"
+
+	"github.com/optakt/flow-dps/rosetta/failure"
+	"github.com/optakt/flow-dps/rosetta/identifier"
+	"github.com/optakt/flow-dps/rosetta/object"
+)
+
+// CombineRequest implements the request schema for /construction/combine.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#request
+type CombineRequest struct {
+	NetworkID           identifier.Network `json:"network_identifier"`
+	UnsignedTransaction string             `json:"unsigned_transaction"`
+	Signatures          []object.Signature `json:"signatures"`
+}
+
+// CombineResponse implements the response schema for /construction/combine.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#response
+type CombineResponse struct {
+	SignedTransaction string `json:"signed_transaction"`
+}
+
+// Combine implements the /construction/combine endpoint of the Rosetta Construction API.
+// It creates a signed transaction by combining an unsigned transaction and
+// a list of signatures.
+// See https://www.rosetta-api.org/docs/ConstructionApi.html#constructioncombine
+func (c *Construction) Combine(ctx echo.Context) error {
+
+	var req CombineRequest
+	err := ctx.Bind(&req)
+	if err != nil {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidEncoding(invalidJSON, err))
+	}
+
+	if req.NetworkID.Blockchain == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(blockchainEmpty))
+	}
+	if req.NetworkID.Network == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(networkEmpty))
+	}
+
+	if req.UnsignedTransaction == "" {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(txBodyEmpty))
+	}
+
+	if len(req.Signatures) != 1 {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(signatureListInvalid))
+	}
+
+	signature := req.Signatures[0]
+	if signature.SignatureType != FlowSignatureAlgorithm {
+		return echo.NewHTTPError(http.StatusBadRequest, invalidFormat(txSignatureInvalid))
+	}
```

Comment:
Why are we given an array of signatures and yet we only use the first one and discard the rest?

Reply:
Rosetta API defines the request format, which uses an array of signatures.

Since Rosetta does not support multi-stage signing, we only support a single signer - the account sending the funds.

It's possible that an account can have multiple keys with varying weights, and that a combination of keys is needed. In that case we'd need multiple signatures from a single account.

However, we talked about not supporting this for the time being. So we only expect a single signature.